src/retry-queue.js

```diff
@@ -0,0 +1,104 @@
+import { RequestQueueScaffold } from './base-request-queue'
+import { encodePostData, xhr } from './send-request'
+
+export class RetryQueue extends RequestQueueScaffold {
+    constructor(captureMetrics) {
+        super()
+        this.captureMetrics = captureMetrics
+        this._requestRetriesMap = {} // <RequestId, number>
+        this._counterToQueueMap = {}
+        this._pollerCounter = 1
+        this._areWeOnline = 'onLine' in window.navigator ? window.navigator.onLine : true
+        this._offlineBacklog = []
+
+        if ('onLine' in window.navigator) {
+            window.addEventListener('online', () => {
+                this._handleWeAreNowOnline()
+            })
+            window.addEventListener('offline', () => {
+                this._areWeOnline = false
+            })
+        }
+    }
+
+    enqueue(requestData) {
+        const { requestId } = requestData
+        let retriesPerformedSoFar = 0
+        if (!(requestId in this._requestRetriesMap)) {
+            this._requestRetriesMap[requestId] = 0
+        } else {
+            retriesPerformedSoFar = this._requestRetriesMap[requestId]
+            if (retriesPerformedSoFar === 10) {
+                delete this._requestRetriesMap[requestId]
+                return
+            }
+            this._requestRetriesMap[requestId]++
+        }
+
+        const nextRetry = this._pollerCounter - 1 + 2 ** retriesPerformedSoFar
+        if (!(nextRetry in this._counterToQueueMap)) {
+            this._counterToQueueMap[nextRetry] = []
+        }
+        this._counterToQueueMap[nextRetry].push(requestData)
+        this.poll()
+    }
+
+    poll() {
+        clearTimeout(this._poller)
+        this._poller = setTimeout(() => {
```

Comment:
I think it's valid to use `setInterval` actually. The reason we were using `setTimeout` is because that's what the other queue uses so that it can stop polling if the user goes idle. This queue used to do the same but doesn't anymore, so should be fine.

Reply:
So I went back to read what I about what I thought was the case (`setInterval` is evil :imp: ) and decided against using it. You did point out a serious flaw about the polling here - that `enqueue` would clear the timeout in the middle if an event came in. This was because of some logic I had removed here. 

I've updated it so that the first enqueue call will start the poller, which will then run via `setTimeout` triggering a new poll.

Reply:
We definitely should not use `setInterval`. It's evil like @yakkomajuri said. Too many ways it can bite you... and the interval duration nor reliability itself is not guaranteed either.

Reply:
.. ok wow, just read about this, and I'm shocked how bad it is. 😮